### ISSUE TITLE
DMI_DOH throws false positives

### ISSUE THREAD

**Issue by hgschmie** [spotbugs/spotbugs#1300]
consider this class using google guava:

```java
package bugs;

import com.google.common.base.Strings;


public final class BugTest {
    String test = "Hello, World";

    public boolean isTestNull() {
        return Strings.nullToEmpty(test).trim().isEmpty();
    }
}
```

with this main class:

```
package bugs;

public class Main {
    public static final void main(String ... args) {
        BugTest bugTest = new BugTest();
        System.out.println(bugTest.isTestNull());
    }
}
```

Spotbugs reports:

```
[INFO] BugInstance size is 1
[INFO] Error size is 0
[INFO] Total bugs: 1
[ERROR] Medium: D'oh! A nonsensical invocation of nullToEmpty(String) in bugs.BugTest.isTestNull() [bugs.BugTest] At BugTest.java:[line 10] DMI_DOH
```

Consider this main method (which can be in another jar and compiled independently from the BugTest class):

```
package bugs;

public class Main {
    public static final void main(String ... args) {
        BugTest bugTest = new BugTest();
        System.out.println(bugTest.isTestNull());
        bugTest.test = null;
        System.out.println(bugTest.isTestNull());
    }
}
```

running it returns correctly:

```
false
true
```

(and using this main class in the same jar as BugTest also makes the error message disappear)

Spotbugs somehow assumes that a package-private field can not be modified outside the compilation unit, so it reports a DMI_DOH if the field is not touched within the compilation unit.

**Comment by hgschmie** [comment]
Spotbugs 4.0.4 using OpenJDK 11.0.8 (from Fedora Linux 32) using the maven spotbugs plugin.

**Comment by ThrawnCA** [comment]
@hgschmie Modifying a package private field does seem like a poor practice.

**Comment by hgschmie** [comment]
I guess the job here is not to judge whether something is what one of us would write or not but whether it is causing a bug or not. While I would agree that this is not the style that I would write code in, it is IMHO a false warning because the checker believes the field to be not accessible (this is why it is reported) while in reality, it is accessible.

I leave the debate around "you should not do that" to others.
